ros/src/computing/planning/motion/packages/waypoint_follower/nodes/mpc_follower/mpc_waypoints_converter.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2018-2019 Autoware Foundation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <ros/ros.h>
+#include <std_msgs/Int32.h>
+#include <autoware_msgs/Lane.h>
+#include <autoware_msgs/VehicleStatus.h>
+
+class MPCWaypointsConverter
+{
+public:
+  MPCWaypointsConverter()
+  {
+    pub_waypoints_ = nh_.advertise<autoware_msgs::Lane>("/mpc_waypoints", 1);
+    sub_closest_waypoint_ = nh_.subscribe("/closest_waypoint", 1, &MPCWaypointsConverter::callbackClosestWaypoints, this);
+    sub_base_waypoints_ = nh_.subscribe("/base_waypoints", 1, &MPCWaypointsConverter::callbackBaseWaypoints, this);
+    sub_final_waypoints_ = nh_.subscribe("/final_waypoints", 1, &MPCWaypointsConverter::callbackFinalWaypoints, this);
+
+    closest_idx_ = 0;
+    back_waypoints_num_ = 10;
+    front_waypoints_num_ = 50;
+  };
+  ~MPCWaypointsConverter(){};
+
+private:
+  ros::NodeHandle nh_, pnh_;
+  ros::Publisher pub_waypoints_;
+  ros::Subscriber sub_closest_waypoint_, sub_base_waypoints_, sub_final_waypoints_, sub_current_velocity_;
+
+  autoware_msgs::Lane base_waypoints_;
+  int closest_idx_;
+  int back_waypoints_num_;
+  int front_waypoints_num_;
+
+  void callbackClosestWaypoints(const std_msgs::Int32 msg)
+  {
+    closest_idx_ = msg.data;
+  }
+
+  void callbackBaseWaypoints(const autoware_msgs::Lane &msg)
+  {
+    base_waypoints_ = msg;
+  }
+
+  void callbackFinalWaypoints(const autoware_msgs::Lane &final_waypoints)
+  {
+    if (base_waypoints_.waypoints.size() == 0 || final_waypoints.waypoints.size() == 0)
+      return;
+
+    if ((int)base_waypoints_.waypoints.size() - 1 < closest_idx_)
+    {
+      ROS_WARN("base_waypoints_.waypoints.size() - 1 = %d, closest_idx_ = %d", (int)base_waypoints_.waypoints.size(), closest_idx_);
+      return;
+    }
+
+    auto sq_dist = [](const geometry_msgs::Point &a, const geometry_msgs::Point &b) {
```

Comment:
Please use amathutils implementation and square the return

Reply:
The reason for using `sq_dist` here is to omit the computation of sqrt. This process is expensive when the size of the waypoint grows.